content/en/blog/basics-for-passwordless/index.md

```diff
@@ -0,0 +1,136 @@
+The core functionality consists of two features: registering a new user and authenticating
+an existing user (and returning an access token after a successful authentication). Of course,
+an end-product would have more features, e.g., for adding and removing authenticators
+but starting with the core features is the simplest.
+
+We wrote the authentication server in Go.
+It utilizes [`go-webauthn`](https://github.com/go-webauthn/webauthn) library.
+Both of the core features need two API endpoints.
+On the client side, we use
+the [navigator credentials JavaScript API](https://w3c.github.io/webappsec-credential-management/#framework-credential-management)
+in a React application. The following sequence graphs demonstrate in more detail how the logic flows:
+
+**Registration**
+
+```mermaid
+sequenceDiagram
+    autonumber
+    participant Client
+    participant Server
+    Client->>Server: Request for credential creation options.<br/>/attestation/options
+    Server-->>Client: Return credential creation options.
+    Note left of Client: Create credential with received options.<br/>navigator.credentials.create
+    Client->>Server: Send client data and public key to server.<br/>/attestation/result
+    Note right of Server: Validate data. <br/> Store the public key and the credential id, </br> and attach to user.
+    Server-->>Client: Registration OK!
+```
+
+**Authentication**
+
+```mermaid
+sequenceDiagram
+    autonumber
+    participant Client
+    participant Server
+    Client->>Server: Request for credential request options.<br/>/assertion/options
+    Server-->>Client: Return credential request options.
+    Note left of Client: Get credential with received options. <br/> Create signature with the private key.<br/>navigator.credentials.get
+    Client->>Server: Send assertion to server.<br/>/assertion/result
+    Note right of Server: Validate data and verify signature.
+    Server-->>Client: Return access token.
+```
+
+For more information on how and which data is handled, see, for example, [the WebAuthn guide](https://webauthn.guide/#webauthn-api).
+You can also find our [client application source codes](https://github.com/findy-network/findy-wallet-pwa/blob/master/src/components/WebauthnLogin.tsx)
+and [the authentication service implementation](https://github.com/findy-network/findy-agent-auth/blob/master/main.go)
+on GitHub.
+
+As the example above shows, implementing a passwordless is not impossible.
+However, as with any new technology, it takes time and dedication from the team
+to familiarize themselves with the new logic and concepts. In addition,
+as there are strict security requirements, testing the authentication flows
+in a local development environment might be challenging at times, as the client and server
+need to belong to the same domain. The local testing environment is something that the team
+should resolve and enable together already early in the project so that it will not become a bottleneck.
+
```

Comment:
I mentioned the architecture 'thing' earlier. As far as I understand web apps, I have always thought pros should handle authentication. That leads to preferring federated authentication or OIDC, at least. What kind of implications could the security (end decentralization) passkey have? Storing just public keys is far better than storing password hashes.

I just wanted to let you know that I leave it to you to think if you want to bring these things to the blog or if they are even relevant.

Reply:
Yes, it would be good to sum the technology pros but I am thinking this post already the next step: the reader has heard about this new passwordless technology, has understood the benefits and now wants to take it into use.

It is true that many apps do and should do the authentication in the way you mentioned, but in many applications the first implemented authentication method is username+password. Also, I think that it would be good to play around with this tech just to undestand the basics better.

Good comments, and they make me start to wonder if I should publish this at all 😄